### PR TITLE
[Python] Append columns

### PR DESCRIPTION
**Pull request by eddyxu**
Provide equivalent of C++ API (#299) in Python.

### CODE REVIEW

File: python/lance/_lib.pyx [hunk 1]
```diff
@@ -138,6 +146,56 @@ cdef class LanceFileFormat(FileFormat):
...
+        with nogil:
+            check_status(self.sp_updater.get().UpdateBatch(arr))
+
+    def finish(self):
+        cdef shared_ptr[CLanceDataset] new_dataset = GetResultValue(
+            self.sp_updater.get().Finish()
+        )
+        return FileSystemDataset.wrap(static_pointer_cast[CDataset, CLanceDataset](new_dataset))
+
+    def __iter__(self):
```
Comment (changhiskhan) [inline python/lance/_lib.pyx:189]: Why do you need both __iter__ and next?

File: python/lance/_lib.pyx [hunk 2]
```diff
@@ -220,6 +279,28 @@ cdef class FileSystemDataset(Dataset):
         c_version = GetResultValue(self.lance_dataset.latest_version())
         return _dataset_version_to_json(c_version)
 
+    def append_column(self, field: Field, func: Callable[[pyarrow.Table], pyarrow.Array]) -> FileSystemDataset:
```
Comment (changhiskhan) [inline python/lance/_lib.pyx:282]: User needs to know that func input is just a chunk of data. Eg they can't do something like normalization in this function unless they explicitly compute the mean / std globally.

**Review by changhiskhan — APPROVED** [review]
I think docs need to be clearer that the func is chunks. I see there's there's a separate issue for projection that works. Also let's add support for user to compute new column externally and pass it in (if they need to say normalize a column)?

File: python/lance/_lib.pyx (hunk 1 above)
Comment (eddyxu) [reply python/lance/_lib.pyx:189]: Do you mean `next()` or `__next__()` here?

File: python/lance/_lib.pyx (hunk 2 above)
Comment (eddyxu) [reply python/lance/_lib.pyx:282]: k, will do.

**Comment by eddyxu** [comment]
> Also let's add support for user to compute new column externally and pass it in (if they need to say normalize a column)?

Filed https://github.com/eto-ai/lance/issues/320. As discussed offline, the main challenging is make the layout aligned with the existing data with the new column. Need some designs.

File: python/lance/_lib.pyx (hunk 2 above)
Comment (eddyxu) [reply python/lance/_lib.pyx:282]: This is actually a good argument of whether func should take in a RecordBatch instead of Table

**Pull request merged**